### ISSUE TITLE
Perform code bundling asynchronously during preflight execution

### ISSUE THREAD

**Issue by Chriscbr** [winglang/wing#5523]
### Use Case

Improve the time it takes to compile Wing programs for platforms that require inflight code bundles (almost all except the simulator, perhaps)

### Proposed Solution

Improve bundling performance by bundling in some kind of asynchronous fashion. Bundling is fairly I/O intensive, so if parallelized (so multiple bundles are produced at the same time, or so bundles are produced as the same time as unrelated preflight code is executed), it could be significantly faster.

The current synchronous code is here:

https://github.com/winglang/wing/blob/9c3a40d7e51c2963b19d1f7614f9b551cad66bb6/libs/wingsdk/src/shared/bundling.ts#L39

### Implementation Notes

Async/await [isn't supported by the CDK programming model](https://github.com/aws/aws-cdk/issues/8273) so the implementation strategy is unclear 

### Component

Compiler

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.
* If this issue is labeled **needs-discussion**, it means the spec has not been finalized yet. Please reach out on the #dev channel in the [Wing Slack](https://t.winglang.io/slack).

**Comment by MarkMcCulloh** [comment]
> Async/await [isn't supported by the CDK programming model](https://github.com/aws/aws-cdk/issues/8273) so the implementation strategy is unclear

Making App's `synth` and `_preSynthesize` (and all other resource `_preSynthesize`) methods async is the solution, no? With your recent change to do bundling in `_preSynthesize`, only that and `synth` would need to be awaited. Neither of those are called in constructors afaik, so from a technical perspective we should be in the clear. I'd be happy to tackle this approach if there's no resistance.

I think the main fun part is that we'd want to execute `_preSynthesize` in parallel but while still respecting dependencies (i.e. if construct B depends on construct A, then `A._preSynthesize()` should finish before `B._preSynthesize()`). Maybe that part isn't necessary, but I think we can have bundles that depend on other bundles.

Even if there are concerns about this ideologically, the lifecycle of the App is not controllable in wing without creating a new platform, in which case I think the user should be trusted with the responsibility of async. 
`preflight` in wing itself will remain synchronous as far as the user is concerned.

**Comment by eladb** [comment]
I support turning `preSynthesize` to asynchronous (this is basically what we called in the original wing spec "async finalizer") - let's even call this `finalize` and turn it into a non-internal method because it needs to be possible to implement this from user land.

As for kicking off all finalizers in parallel - I don't see an issue because order and dependencies at this stage are not guaranteed already.

**Comment by MarkMcCulloh** [comment]
> Would this mean an end to reproducible, deterministic compilations?

Nope, it would not. I think the argument has been made pretty clear by many in https://github.com/aws/aws-cdk/issues/8273, which is why several cdk-based metaframeworks allow async code in presumably reproducible/deterministic code paths.

async != inherently non-deterministic (at least, not in an inherently different way than sync). It does add a new option for non-determinism, but it's not like it's impossible to be non-deterministic synchronously.

If we want to limit ourselves to perfectly "pure" code paths, we should also disallow any io operations or syscalls during synth.

**Comment by Chriscbr** [comment]
> async != inherently non-deterministic (at least, not in an inherently different way than sync). It does add a new option for non-determinism, but it's not like it's impossible to be non-deterministic synchronously.

Right -- I suppose it just might require more defensive implementations of resources for class/library authors. For example, if some asynchronous work is adding environment variables to a cloud function (where the execution could be reordered), someone needs to make sure all added keys are eventually sorted before they're synthesized into Terraform or other configuration files.

**Comment by MarkMcCulloh** [comment]
Yeah, and this would only really be possible from preSynth hooks which already are not supposed to guarantee an order. Maybe we should expose a different hook that is reliably ordered (and synchronous)?
